File: haskell/Bob.hs
{-# LANGUAGE OverloadedStrings #-}

module Bob (responseFor) where

import qualified Data.Text as T
import Data.Text (Text)
import Data.Char

isQuestion :: Text -> Bool
isQuestion text =
  case T.unsnoc text of
    Nothing -> False
    Just (_, c) -> c == '?'

isAllUpper :: Text -> Bool
isAllUpper text
  | T.null filtered = False
  | otherwise = T.all isUpper filtered
    where
      filtered = T.filter isAlpha text

responseFor :: Text -> Text
responseFor text
    | T.null stripped = "Fine. Be that way!"
    | isAQuestion && isAllUpper stripped
      = "Calm down, I know what I'm doing!"
    | isAQuestion = "Sure."
    | isAllUpper text = "Whoa, chill out!"
    | otherwise = "Whatever."
      where
        stripped = T.strip text
        isAQuestion = isQuestion stripped

File: F#/Bob.fs
module Bob

open System

let private everypred preds = function
    | xs -> Seq.fold (fun r p -> r && p xs) true preds

let silence = Seq.isEmpty

let shouting s =
    match Seq.filter Char.IsLetter s with
    | s when silence s -> false
    | s -> Seq.forall Char.IsUpper s

let question s =
    match Seq.tryLast s with
    | Some c -> c = '?'
    | _ -> false

let forcefulQuestion = everypred [ question; shouting ]

let response (input: string) =
    match input.Trim() with
    | s when forcefulQuestion s -> "Calm down, I know what I'm doing!"
    | s when shouting s -> "Whoa, chill out!"
    | s when question s -> "Sure."
    | s when silence s -> "Fine. Be that way!"
    | _ -> "Whatever."